File: problems/1-easy/20-Valid-Parentheses/solution.cpp
#include <unordered_map>
#include <vector>
#include <string>
using namespace std;

bool isValid(string s) {
    unordered_map<char, char> mapping = {{'(', ')'}, {'[', ']'}, {'{', '}'}};
    vector<char> stack = {};

    for (char c : s) {
        if (mapping.find(c) != mapping.end()) {
            stack.push_back(c);
        } else {
            if ((stack.size() > 0) && mapping[stack.back()] == c) {
                stack.pop_back();
            } else {
                return false;
            }
        }
    }

    return stack.size() == 0;
}

File: problems/1-easy/20-Valid-Parentheses/solution.py
def isValid(s: str) -> bool:
    my_dict = {")": "(", "]": "[", "}": "{"}
    stack = []

    for char in s:
        if char not in my_dict:
            stack.append(char)
        elif stack and my_dict[char] == stack[-1]:
            stack.pop()
        else:
            return False

    return len(stack) == 0
